##### Question 

#### Solution Part A:
We need find the vlaue of z (from the Gambler's Ruin problem ) where Pz is the probability of reaching M= 8 wiothut reaching 0. here : $$ 
p (\textrm{Prbability of wining} ) = p = .4\\
q (\textrm{Probablity of Lossing}) =1-p=q=.6 \\
z = 1 \\
M = 8 \\
\frac{q}{p} = \frac{0.6}{0.4} = 1.5 $$

$$ P = \frac{1-(\frac{q}{p})^z}{1-(\frac{q}{p})^M} \\
P = \frac{1-1.5^{1}}{1-1.5^{8}} $$

In [7]:
import numpy as np

p = 0.4
q = 0.6
qbyp = q/p

p8 = round((1-qbyp**1)/(1-qbyp**8),4)

print("Probability of reaching $8 without going to $0 is:",p8)


Probability of reaching $8 without going to $0 is: 0.0203


(b) he bets, each time, as much as possible but not more than necessary to bring his fortune up to 8 dollars (bold strategy).

Here Smith would spend \$1 and would get \$2 if he wins, and \$4 if he wins again and \$8 if he wins again, asumaing he is betting with all the money he got from each bet. so In this case we need to find how soon we can reach maximum probability with stage of $8.

So if Smith bets with \$2 he will either win \$4 or loose \$2 and reach \$0. Hence all the stage would have probability of.6 for \$0 state.

| Gain | State | 0   | 1   | 2   | 4   | 8   |
|:----:|------|-----|-----|-----|-----|-----|
| 0    | 0    | 1   | 0   | 0   | 0   | 0   |
| 0+1  | 1    | 0.6 | 0   | 0.4 | 0   | 0   |
| 1+1  | 2    | 0.6   | 0 | 0   | 0.4 | 0   |
| 2+2  | 4    | 0.6   | 0   | 0 | 0   | 0.4 |
| 4+4  | 8    | 0   | 0   | 0   | 0   | 1   |

starting state vector = v= [ 0,1,0,0,0]

In [10]:
import numpy as np

# Probability Transition matrix
p = np.array([[1,0,0,0,0],
              [0.6,0,0.4,0,0],
              [0.6,0,0,0.4,0],
              [0.6,0,0,0,0.4],
              [0,0,0,0,1]])

#stage zero vector, or starting stage 
Sv = np.array([[0,1,0,0,0]])

In [11]:
# Stage p1 after 1st Transition 
p1 = Sv @ p
p1

array([[0.6, 0. , 0.4, 0. , 0. ]])

Result shows that at this point we have 0.6 probability of going to stage \$0 and 0.4 to goto \$2 stage.

In [12]:
p2 = p1 @ p
p2

array([[0.84, 0.  , 0.  , 0.16, 0.  ]])

Result shows that at this point we have 0.84 probability of going to stage \$0 and 0.16 to goto \$4 stage.

In [13]:
p3= p2 @ p
p3

array([[0.936, 0.   , 0.   , 0.   , 0.064]])

Result shows that at this point we have 0.936 probability of going to stage \$0 and 0.064 to goto \$8 stage.

In [16]:
p4= p3 @ p
p5 = p4 @ p
p4,p5

(array([[0.936, 0.   , 0.   , 0.   , 0.064]]),
 array([[0.936, 0.   , 0.   , 0.   , 0.064]]))

here we have reached the stationary stage of  transtion matrix, and we are always getting probability of .064 for getting \$8 without going \$0.

(c) Which strategy gives Smith the better chance of getting out of jail?`

Probability with Timid strategy :  0.0203
Probability with Bold strategy : 0.064
This is very clear that Bold strategy gives more than three times probability of winning the bet. 


Using Binomial probability checking answer of Bold Strategy .

In [39]:
from scipy.stats import geom,binom,poisson
import math 

In [40]:
# Probability Statement:
# P(X=0 success in 8 yrs)

n = 2 
p = .4 # Probabaility of Failure 
q = 1-p # Probability of success 
k = 2

bp2 = binom.pmf(k,n,p)
bp3 = binom.pmf(k+1,n+1,p)
bp4 = binom.pmf(k+2,n+2,p)

print("probability that  Smith will win after 2 bets:",round(bp2,3))
print("Expected value :", round(n*p,3))
print("probability that  Smith will win after 3 bets:",round(bp3,3))
print("Expected value :", round(((n+1)*p),3))



probability that  Smith will win after 2 bets: 0.16
Expected value : 0.8
probability that  Smith will win after 3 bets: 0.064
Expected value : 1.2
